Add .random method to MvGaussianRandomWalk (Issue #4337)

awray3 opened the pull request:
This PR addresses Issue #4337. So far I have implemented the .random method for `MvGaussianRandomWalk` and updated a few of the docstrings in this file. The implementation is similar to `GaussianRandomWalk.random`.

Currently I am having issues with adding this method to `test_distributions_random.py`; I get many failing tests and I am not sure if the base test class is appropriate with its current settings for this time series.

awray3 committed:
Implement random for MvGaussianRandomWalk

Implements the random method for MvGaussianRandomWalk,
partially fixing #4337.

awray3 committed:
Update docstring for GaussianRandomWalk._random()

awray3 committed:
Add example to random docstring

Sayam753 commented on pymc3/distributions/timeseries.py:
```diff
@@ -464,6 +464,86 @@ def logp(self, x):
+            },
+        )
+
+    def _random(self, mu, param, size, sample_shape, cov_type):
```
```suggestion
multivariate_samples = self.innov.random(point=point, size=size)
```
How about the use of MvNormal random method here?
Because if `cov` is a random variable, then drawing samples from it can have more than 2 dimensions and scipy.stats.multivariate_normal cannot handle a batch of multivariate normals.

awray3 replied:
That sounds like a great idea. What is the shape of the output of `MvNormal`'s random array? I'm trying to think of a good way to select out the time axis to perform the cumulative sum.

Sayam753 replied:
> What is the shape of the output of MvNormal's random array

If you pass (time_dims, n_events) shape to MvNormal, then output shape will be (sample_shape, time_dims, n_events).

Sayam753 commented on pymc3/distributions/timeseries.py:
```diff
@@ -464,6 +464,86 @@ def logp(self, x):
+
+        # time axis comes after the sample axis
+        time_axis = len(sample_shape)
+
```
What if sample shape is None? Then time_axis = 0. 
Is this correct behaviour?

awray3 replied:
Yes, that was my intent in my version of `_random`. My understanding was that the `_random` function will receive sample size of None when the user calls `MvGaussianRandomWalk.random()` without specifying size. In this case I think the shape of the random walk will be (timesteps, spatial_dims), so having time_axis = 0 would be right.

Sayam753 replied:
Makes sense 👍

Sayam753 reviewed (changes requested):
Looks like a great start @awray3 . I addressed a few things below.
Let's iterate over it and get it ready for merge.

Sayam753 commented:
> I get many failing tests

Do you get failing tests in existing test cases?

awray3 commented:
> > I get many failing tests
> 
> Do you get failing tests in existing test cases?

No, just for new tests created when I include a new test called `TestMvGaussianRandomWalk`.

awray3 committed:
Modify MvGaussianRandomWalk.random

Improves the implementation by using MvNormal.random as suggested by
@Sayam753. Also updated its docstring to add more examples.

awray3 committed:
Add TestMvGaussianRandomWalk

awray3 commented on pymc3/distributions/timeseries.py:
```diff
@@ -439,7 +439,7 @@ def __init__(
         self.init = init
         self.innovArgs = (mu, cov, tau, chol, lower)
-        self.innov = multivariate.MvNormal.dist(*self.innovArgs)
+        self.innov = multivariate.MvNormal.dist(*self.innovArgs, shape=self.shape[-1])
```
The idea here is to pass the spatial dimension to the `MvNormal` function. I think there might be a better way to do this.

Sayam753 commented on pymc3/distributions/timeseries.py:
```diff
@@ -464,6 +464,69 @@ def logp(self, x):
+        # samples from MvNormal.
+        size_time_steps = size + to_tuple(time_steps)
+
+        multivariate_samples = self.innov.random(point=point, size=size_time_steps)
```
So, you are passing spatial dims during initialization of MvNormal and pass the `time_dims` along with `size` here while drawing samples. 

Let's say I want to model just a single time series and passed in the shape just (spatial_dims) instead of (1, spatial_dims) to `MvGaussianRandomWalk`. Then with the shape handing above, multivariate samples will be drawn with shape (sample_shape, spatial_dims, spatial_dims).

I would say better it would be to pass 
1. (time_dims, spatial_dims) during initialization of MvNormal.
2. only (size) while drawing samples to take care of corner cases for us.

awray3 replied:
Great point. I've made some more changes that I will push soon. I'm finding an error being thrown from `MvGaussianRandomWalk.logp` when I use `shape = 2` instead of `shape = (1, 2)`. Is passing just an int (or None for that matter) for shape supported by `MvGaussianRandomWalk`?

Sayam753 replied:
Looking at the logp code, I think it only supports two dimensional shape as input.
How about adding something similar as MvNormal logp? as the latter handles single dimensional shape as well.

https://github.com/pymc-devs/pymc3/blob/3cfee77fda041483e86d9c61fa1d1ada9380cdb3/pymc3/distributions/multivariate.py#L114-L116

Sayam753 replied:
> None for that matter

This is an interesting case. MvNormal expects any of cov, chol or tau to be a square matrix. So, I expect to specify at least one dimension when initialising MvNormal. Same reasoning applies to MvGaussianRandomWalk as well.

awray3 replied:
Okay, I think that worked! Thanks for the help, @Sayam753. The current functionality when multiple time steps are requested is to center the walk so that it starts at 0. Should this also be done when no time dimension is passed in? It would seem silly to write it that way, since then passing something like `shape = 2` would always give you an array of zeros. My guess is that if shape has only one dimension, then the user wants a sample from a multivariate normal, so that's what should actually get returned.

(I suppose if the user really wants a bunch of arrays of zeros they could always just pass `shape = (1, 2)` !)

Sayam753 replied:
Good point. Let's not center the multivariate samples at zero in that case. If I understood correctly, the cumulative sum operation is performed irrespective whether there is a single dimension or twice?

awray3 replied:
Well, the way I rewrote it (which I haven't pushed yet) only computes a cumulative sum if there is a time axis given as well. Otherwise it skips that step too and just returns the normal samples.

awray3 commented:
How should `MvGaussianRandomWalk` be added to the testing scripts? I think the way I added it in doesn't make much sense with many of the shapes supplied by `BaseTestCases`.

awray3 committed:
Pass entire shape to MvNormal

This commit rewrites the random method to pass the entire shape
into MvNormal. MvGaussianRandomWalk also now supports an integer shape
that must match the dimensionality of mu. In this case it is assumed
that there are no time steps, and a random sample from the multivariate
normal distribution will be returned.

Sayam753 commented:
> I think the way I added it in doesn't make much sense with many of the shapes supplied by BaseTestCases.

Yes. The `BaseTestCases` class have parametrized shape as an argument that gets passed to the distribution and that can have more than 2 dimensions as well. 

You can add tests very similar to how MvNormal random method is tested with np arrays here -
https://github.com/pymc-devs/pymc3/blob/3cfee77fda041483e86d9c61fa1d1ada9380cdb3/pymc3/tests/test_distributions_random.py#L1596-L1605

If this works, then we can add a test where a parameter can be a random variable.

Sayam753 commented on pymc3/distributions/timeseries.py:
```diff
@@ -464,6 +468,74 @@ def logp(self, x):
+        multivariate_samples = self.innov.random(point=point, size=size)
+        # this has shape (size, self.shape)
+
+        if len(self.shape) >= 2:
```
We cannot have more than 2 dimensions for self.shape because Theano does not support `logp` calculations and cholesky decomposition with batches of covariances matrices. 

```suggestion
        if len(self.shape) == 2
```

awray3 replied:
Thanks, @Sayam753! I really like how nice the code is turning out. I'll try to work on getting the tests for RV `mu` and others. I suspect it will need a slight modification of `generate_shapes`.

Sayam753 commented on pymc3/distributions/timeseries.py:
```diff
@@ -464,6 +468,74 @@ def logp(self, x):
+            time_axis = len(size)
+            multivariate_samples = multivariate_samples.cumsum(axis=time_axis)
+
+            if time_axis:
```
```suggestion
            if time_axis != 0:
```
Let's be more explicit here for readability

Sayam753 commented on pymc3/distributions/timeseries.py:
```diff
@@ -464,6 +468,74 @@ def logp(self, x):
+                # size was passed as None
+                multivariate_samples = multivariate_samples - multivariate_samples[0]
+
+        # if the above if statement fails, then only a spatial dimension was passed in for self.shape.
```
```suggestion
        # if the above statement fails, then only a spatial dimension was passed in for self.shape.
```

Sayam753 reviewed (changes requested):
That's a great progress @awray3 . Minor nitpicks below -

Also, the test cases pass with numpy arrays 🎉 . Last step would be to add a test where `mu` and any one of `chol`, `cov` or `tau` are random variables. Similar to how MvNormal is tested.

awray3 committed:
Update pymc3/distributions/timeseries.py

Remove equality since the shape parameter cannot have more than 2 dimensions.

Co-authored-by: Sayam Kumar <sayamkumar753@yahoo.in>

awray3 committed:
Update comment pymc3/distributions/timeseries.py

Co-authored-by: Sayam Kumar <sayamkumar753@yahoo.in>

awray3 committed:
Update pymc3/distributions/timeseries.py

Adds a more explicit conditional on the time_axis statement.

Co-authored-by: Sayam Kumar <sayamkumar753@yahoo.in>

Sayam753 commented:
There is a conflicting file. Can you rebase to current master to start the GitHub Actions?

awray3 commented:
No problem, I just pushed after rebasing. Currently working out an error from `test_with_chol_rv` when `mu` and others are RVs and `dist_shape = (10, 3)` and `mu_shape = (10, 3)`. Looks like another bug with indexing inside of `MvGaussianRandomWalk.logp`. Below is the traceback; I would welcome any pointers in troubleshooting it.

```
...
   test_with_chol_rv(sample_shape=None, dist_shape=(10, 3), mu_shape=(10, 3))
  File "my_test.py", line 31, in test_with_chol_rv
    mv = pm.MvGaussianRandomWalk("mv", mu, chol=chol, shape=dist_shape)
  File "/home/jovyan/pymc3/distributions/distribution.py", line 121, in __new__
    return model.Var(name, dist, data, total_size, dims=dims)
  File "/home/jovyan/pymc3/model.py", line 1135, in Var
    var = FreeRV(name=name, distribution=dist, total_size=total_size, model=self)
  File "/home/jovyan/pymc3/model.py", line 1668, in __init__
    self.logp_elemwiset = distribution.logp(self)
  File "/home/jovyan/pymc3/distributions/timeseries.py", line 466, in logp
    return self.init.logp_sum(x[0]) + self.innov.logp_sum(x_i - x_im1)
  File "/home/jovyan/pymc3/distributions/distribution.py", line 266, in logp_sum
    return tt.sum(self.logp(*args, **kwargs))
  File "/home/jovyan/pymc3/distributions/multivariate.py", line 311, in logp
    quaddist, logdet, ok = self._quaddist(value)
  File "/home/jovyan/pymc3/distributions/multivariate.py", line 127, in _quaddist
    delta = value - mu
  File "/opt/conda/envs/pymc3-dev-py37/lib/python3.7/site-packages/theano/tensor/var.py", line 121, in __sub__
    return theano.tensor.basic.sub(self, other)
  File "/opt/conda/envs/pymc3-dev-py37/lib/python3.7/site-packages/theano/gof/op.py", line 642, in __call__
    compute_test_value(node)
  File "/opt/conda/envs/pymc3-dev-py37/lib/python3.7/site-packages/theano/gof/op.py", line 111, in compute_test_value
    required = thunk()
  File "/opt/conda/envs/pymc3-dev-py37/lib/python3.7/site-packages/theano/gof/op.py", line 880, in rval
    thunk()
  File "/opt/conda/envs/pymc3-dev-py37/lib/python3.7/site-packages/theano/gof/cc.py", line 1848, in __call__
    raise exc_value.with_traceback(exc_trace)
ValueError: Input dimension mis-match. (input[0].shape[0] = 9, input[1].shape[0] = 10)
```

Sayam753 commented:
Oh. This is very much similar to bug described in #4010 . 
Random walk has some weird shape handling for logp when it comes to 2 dimensional inputs.

Sayam753 commented:
Let's think of the case, dist_shape = (10, 3) and mu_shape = (10, 3)
10 time steps and 3 time series to model.

Let's consider an arbitrary tensor with shape (10, 3) following MvGaussianRandomWalk. Its first row (or first time step) comes from init distribution. And that initial distribution is Flat one in our case. The remaining 9 steps will come from adding (9, 3) multivariate samples as noise to the initial row. 

I think the error is because we initialize MvNormal with shape (10, 3) and ask for logp from MvNormal with (9, 3) shaped tensor. And there is a mismatch.

Sayam753 commented:
A good solution for me will be to open an issue for logp miscalculations. Write the failing test cases here and mark them as [xfail](https://docs.pytest.org/en/latest/skipping.html#xfail-mark-test-functions-as-expected-to-fail). 

Another thing I notice, is that after the rebase, this PR contains all the changes from #4368 . But #4368 is now merged into master.  Can you lookup what went wrong in the rebase?

awray3 commented:
Yes, I can look up the issue. I'm new to a lot of the git workflow, so I might have messed something up 😅

awray3 committed:
Fix rebase issue

awray3 commented:
Okay, I royally messed something up (probably by not rebasing as often I should have), but I think I resolved my rebase issue.

awray3 commented:
Actually, I'm still not sure I fixed my rebase issue, and it's causing a headache! I need to step away and come back with fresh eyes.

Sayam753 commented:
Oh. This is getting mixed with commits. Can you give this a try?

```bash
git checkout 79e47b19  # Let's go to a cleaner commit
git checkout -b add_random_2
git push -f origin add_random
```

Another option can be resetting back and force push
```bash
git reset --hard 3200a84a
git push -f origin add_random
```

awray3 committed:
Fix syntax error

twiecki commented on pymc3/distributions/timeseries.py:
```diff
@@ -275,7 +275,7 @@ def _random(self, sigma, mu, size, sample_shape):
         axis = len(size) - 1 denotes the axis along which cumulative sum would be calculated.
         This might need to be corrected in future when issue #4010 is fixed.
-        Lines 318-322 ties the starting point of each instance of random walk to 0"
+        Lines 291-295 set the starting point of each instance of the random walk to 0."
```
I would refer to the method name instead of lines which is going to be brittle.

awray3 replied:
Good idea; I moved the comment from the docstring to the location in the code in 1761a2e.

twiecki commented on pymc3/distributions/timeseries.py:
```diff
@@ -460,6 +464,72 @@ def logp(self, x):
+        -------
+        Create one sample from a 2-dimensional Gaussian random walk with 10 timesteps::
+
+            mu = np.array([1.0, 0.0])
```
There's specific RST syntax for python code you should use. You can find that in other parts of the code base.

awray3 replied:
I'm comparing my examples' formatting to `MvNormal`'s docstring's examples, and my formatting looks identical to me. Should I be looking elsewhere?

awray3 committed:
Add tests for MvGaussianRandomWalk with RV params

Added tests for both chol and cov to be random variables.

awray3 committed:
Improve comment clarity in MvNormal._random

Moves the brittle comment in MvRandom._random's docstring to the actual location in the code for clarity.

awray3 commented:
> Another option can be resetting back and force push
> 
> ```shell
> git reset --hard 3200a84a
> git push -f origin add_random
> ```

That definitely worked. I appreciate the lending of git expertise! I definitely have more tricks to learn. 🙂

twiecki commented:
@Sayam753 Feel free to merge when you think this is ready.

awray3 commented:
Hooray! First pull request to be merged! 🥳  I also want to implement random methods for `MvStudentsTRandomWalk` and others listed in #4337. Should I open a new pull request for those when I finish them?

Also, I notice that `MvNormal`, `MvGaussianRandomWalk`, and `MvStudentsTRandomWalk` will likely all be tested in similar ways. Would it be sensible to refactor some of the tests to have a shared base class, such as `BaseMvTests`?

Sayam753 commented:
`MvStudentTRandomWalk` random walk is inherited from `MvGaussianRandomWalk`. So, it reuses the random method implemented in this PR. We need to make sure the entire `self.shape` is passed during initialization.

I had not paid attention to the reason behind starting the random draws from zero. But looking at the https://github.com/pymc-devs/pymc3/pull/3985#discussion_r448385963 , I am thinking this way -
1. When we have 2 dimensional input shape, we need to make first row to all zeros (making all time series start at 0). This PR includes that.
2. When we have 1 dimensional input shape, we need to make first element of each row to zero, thereby matching the random draws from `init` distribution. This PR does not handle that. We were thinking doing so, will make all random draws to zero. 
3. On top of it, a user can pass any `init` distribution. So, the random methods have to adapt initial values from that distribution. Feel free to ask any questions on these.

> Would it be sensible to refactor some of the tests to have a shared base class, such as `BaseMvTests`?

That's a good point.

For the next PR, I have these ideas -
 - Making sure `MvStudentTRandomWalk` is correctly initialized.
 - Designing a case class `BaseMvTests` for all the similar test cases.
 - Refactoring the random methods of `GaussianRandomWalk` and `MvGaussianRandomWalk` to start from random values drawn from initial distribution.
Let's polish out all RandomWalk `random` methods before going on to AutoRegressive random methods.

Sayam753 commented:
Last step would be wrapping the docstring examples in `code-block` directive (similar to Normal distribution example [here](https://github.com/pymc-devs/pymc3/blob/e2ce81576ccdc745915bd74b640a3d10d1783082/pymc3/distributions/continuous.py#L469-L475)) and also a mention in RELEASE-NOTES.md. Then it will be ready :)

awray3 committed:
Update docstring formatting

awray3 committed:
Update RELEASE-NOTES.md

awray3 commented:
> For the next PR, I have these ideas

These all sound like great ideas. I also like the idea of polishing the random walks before moving on.

Sayam753 commented:
Thanks @awray3 and congrats on your first Pull Request 🎉

awray3 commented:
Thank you @Sayam753 and @twiecki for helping me get up to speed! I appreciate all the help.
